File: LeetCode/src/main/java/MergeSort.java
import java.util.Arrays;

public class MergeSort {

  public static void main(String[] args) {
    int[] arr = {3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5};
    System.out.println("Original array: " + Arrays.toString(arr));

    int[] sortedArr = mergeSort(arr);

    System.out.println("Sorted array: " + Arrays.toString(sortedArr));
  }

  public static int[] mergeSort(int[] arr) {
    if (arr.length <= 1) {
      return arr;
    }

    int mid = arr.length / 2;
    int[] left = Arrays.copyOfRange(arr, 0, mid);
    int[] right = Arrays.copyOfRange(arr, mid, arr.length);

    return merge(mergeSort(left), mergeSort(right));
  }

  public static int[] merge(int[] left, int[] right) {
    int[] result = new int[left.length + right.length];
    int i = 0, j = 0, k = 0;

    while (i < left.length && j < right.length) {
      if (left[i] <= right[j]) {
        result[k++] = left[i++];
      } else {
        result[k++] = right[j++];
      }
    }

    // 处理剩余元素
    while (i < left.length) {
      result[k++] = left[i++];
    }
    while (j < right.length) {
      result[k++] = right[j++];
    }

    return result;
  }
}

File: LeetCode/src/main/go/mergeSort/mergeSort.go
package main

import (
	"fmt"
)

func mergeSort(arr []int) []int {
	if len(arr) <= 1 {
		return arr
	}

	mid := len(arr) / 2
	left := mergeSort(arr[:mid])
	right := mergeSort(arr[mid:])

	return merge(left, right)
}

func merge(left, right []int) []int {
	result := make([]int, 0, len(left)+len(right))
	i, j := 0, 0

	for i < len(left) && j < len(right) {
		if left[i] <= right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}

	// 处理剩余元素
	result = append(result, left[i:]...)
	result = append(result, right[j:]...)

	return result
}

func main() {
	arr := []int{3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5}
	sortedArr := mergeSort(arr)

	fmt.Println("Original array:", arr)
	fmt.Println("Sorted array:", sortedArr)
}